File: src/step3/ops/scaling.py
# src/step3/ops/scaling.py

import logging

import numpy as np

from src.common.stencil_block import StencilBlock

# Rule 7: Granular Traceability for Numerical Scaling
logger = logging.getLogger("Solver.Ops.Scaling")

def get_dt_over_rho(block: StencilBlock) -> float:
    """
    Returns the scaling factor (dt / rho) for the Predictor and Corrector Steps.
    
    Compliance:
    - Rule 7: Fail-Fast math audit. Guards against vacuum density (rho=0).
    - Rule 4 (SSoT): Derived from immutable block properties.
    """
    # Rule 8: Contract Violation Check (Density cannot be zero or negative)
    if block.rho <= 0:
        logger.critical(
            f"PHYSICS CRASH: Block {block.id} has invalid density (rho={block.rho}). "
            "Vacuum or negative density would cause infinite acceleration."
        )
        raise ValueError(f"Invalid rho={block.rho} in block {block.id}")

    scaling = block.dt / block.rho

    # Forensic Audit
    if not np.isfinite(scaling):
        logger.error(f"PPE MATH ERROR: dt/rho exploded in {block.id} (dt={block.dt}, rho={block.rho})")
        raise ArithmeticError(f"Scaling factor dt/rho is non-finite in block {block.id}")

    return scaling

def get_rho_over_dt(block: StencilBlock) -> float:
    """
    Returns the scaling factor (rho / dt) for the Pressure Poisson Equation.
    
    Compliance:
    - Rule 7: Fail-Fast math audit. Guards against dt=0.
    - Zero-Debt Mandate: Explicit arithmetic; no intermediate caching.
    """
    # Rule 8: Contract Violation Check (Time step cannot be zero)
    if block.dt <= 0:
        logger.critical(
            f"TEMPORAL CRASH: Block {block.id} has invalid time-step (dt={block.dt}). "
            "Zero dt would cause infinite pressure source terms."
        )
        raise ZeroDivisionError(f"Invalid dt={block.dt} in block {block.id}")

    scaling = block.rho / block.dt

    # Forensic Audit
    if not np.isfinite(scaling):
        logger.error(f"PPE MATH ERROR: rho/dt exploded in {block.id} (rho={block.rho}, dt={block.dt})")
        raise ArithmeticError(f"Scaling factor rho/dt is non-finite in block {block.id}")

    return scaling

File: cpp/src/ops/scaling.cpp
/**
 * @file scaling.cpp
 * @brief Implementation of scaling factors for Navier-Stokes solver with execution tracing and comprehensive parameter guards.
 */

#include "scaling.hpp"
#include <cmath>
#include <stdexcept>
#include <iostream>

#ifdef _OPENMP
#include <omp.h>
#endif

namespace navier_stokes_solver {

double get_dt_over_rho(double dt, double rho) {
    // Contract Violation Check: Time-step must be strictly positive
    if (dt <= 0.0) {
        std::cerr << "TEMPORAL CRASH: Invalid time-step (dt=" << dt << "). "
                  << "Zero or negative dt would cause invalid scaling.\n";
        throw std::invalid_argument("Invalid time-step (dt <= 0) in scaling computation.");
    }

    // Contract Violation Check: Density cannot be zero or negative
    if (rho <= 0.0) {
        std::cerr << "PHYSICS CRASH: Invalid density (rho=" << rho << "). "
                  << "Vacuum or negative density would cause infinite acceleration.\n";
        throw std::invalid_argument("Invalid density (rho <= 0) in scaling computation.");
    }

    #ifdef _OPENMP
    int active_threads = omp_get_max_threads();
    #else
    int active_threads = 1;
    #endif

    std::cout << "[THREAD_TRACE] File: scaling.cpp | Function: get_dt_over_rho"
              << " | Active Threads: " << active_threads << "\n";

    double scaling = dt / rho;

    // Forensic Numerical Audit
    if (!std::isfinite(scaling)) {
        std::cerr << "MATH FAILURE [scaling.cpp]: dt/rho exploded (dt=" << dt << ", rho=" << rho << ").\n";
        throw std::runtime_error("Scaling factor dt/rho is non-finite.");
    }

    return scaling;
}

double get_rho_over_dt(double dt, double rho) {
    // Contract Violation Check: Time-step cannot be zero or negative
    if (dt <= 0.0) {
        std::cerr << "TEMPORAL CRASH: Invalid time-step (dt=" << dt << "). "
                  << "Zero or negative dt would cause infinite pressure source terms.\n";
        throw std::invalid_argument("Invalid time-step (dt <= 0) in scaling computation.");
    }

    // Contract Violation Check: Density must be strictly positive
    if (rho <= 0.0) {
        std::cerr << "PHYSICS CRASH: Invalid density (rho=" << rho << "). "
                  << "Vacuum or negative density would invalidate pressure scaling.\n";
        throw std::invalid_argument("Invalid density (rho <= 0) in scaling computation.");
    }

    #ifdef _OPENMP
    int active_threads = omp_get_max_threads();
    #else
    int active_threads = 1;
    #endif

    std::cout << "[THREAD_TRACE] File: scaling.cpp | Function: get_rho_over_dt"
              << " | Active Threads: " << active_threads << "\n";

    double scaling = rho / dt;

    // Forensic Numerical Audit
    if (!std::isfinite(scaling)) {
        std::cerr << "MATH FAILURE [scaling.cpp]: rho/dt exploded (rho=" << rho << ", dt=" << dt << ").\n";
        throw std::runtime_error("Scaling factor rho/dt is non-finite.");
    }

    return scaling;
}

} // namespace navier_stokes_solver
